users/helauth/pipeline.py

```diff
@@ -102,3 +102,66 @@ def update_user_from_profile(request: WSGIRequest, *, user: User | None = None)
         user.date_of_birth = birthday_info["birthday"]
 
     user.save()
+
+
+def migrate_user_from_tunnistamo_to_keycloak(
+    backend: TunnistamoOIDCAuth,
+    request: WSGIRequest,
+    response: OIDCResponse,
+    user: User | None = None,
+    **kwargs: Unpack[ExtraKwargs],
+) -> dict[str, Any]:
+    if user is None:
+        return {"user": user}
+
+    id_token = IDToken.from_string(response["id_token"])
+    if (
+        id_token is not None  # There is an id token.
+        and id_token.iss.endswith("helsinki-tunnistus")  # Issuer is keycloak.
+        and id_token.is_ad_login  # It is an AD login.
+        and user.email not in ("", None, "ANON.ANONYMIZED@anonymized.net")  # User has a valid email.
+    ):
+        migrate_from_tunnistamo_to_keycloak(email=user.email)
+
+    return {"user": user}
+
+
+@SentryLogger.log_if_raises("Keycloak migration: Failed to migrate user from tunnistamo to keycloak")
+def migrate_from_tunnistamo_to_keycloak(*, email: str) -> None:
+    # Look at the two most recent AD users with the same email address (no profile ID).
+    # There can be more, but the most recent one is the new Keycloak user and the other one
+    # is the old Tunnistamo user after the auth-migration (when auth moved to backend), which we want to migrate.
+    users: list[User] = list(User.objects.filter(email=email, profile_id="").order_by("date_joined")[:2])
```

Comment:
Haven't thought this entirely through yet, but `profile_id=""` might not be a sufficient check against some cases where the same email has been used for both profile (suomifi) and AD; i.e. the `old_user` maybe could be a profile one with this info unset (because no recent login)?
If so, then perhaps it can be weeded out by having no staff/superuser flags set nor roles assigned.

Reply:
If the same email has been used for both `AD` and `Suomi.fi` logins, these would have created two different users in our system. See [this `AD` user](https://tilavaraus.hel.fi/admin/users/user/3852/change/) vs. [this `Suomi.fi` user](https://tilavaraus.hel.fi/admin/users/user/5866/change/).  Furthermore, our logging middleware makes sure that `profile_id` is filled for all profile users on each login.

Reply:
Exactly, it creates two users, so then the question is, whether empty profile id is a fool-proof check to ignore the putative profile variants (because it wasn't always saved). Examination of already existing user pairs sharing the same `hel.fi` email turned up just a handful of cases where this wasn't clear, but after inactivating their dubious incarnations (where `date_joined` was more recent), this should be okay. (See further in other comment.)

Reply:
The profile ID filling has been there for at least last September, so I think that all users that still use the system should have that set.